### PR TITLE
EMBED 959 - Implicit OAUTH2 demo

### PR DESCRIPTION
**Pull request by bryans99**
1. Created an Auth component that is responsible for authentication. User can log in anonymously or using Google. When using google, calls out to the google people me API to get information about the user.
2. Added new demo tab to display the results of google sheets query.
3. Refactored data server demo to remove Auth code (now in auth). Error handling moved to a utility.
3. Modified data server to add auth routes and JWT session tokens
a. auth check (GET) - is token valid
b. auth in (POST) - for google logins, verifies access token
c. auth out (GET) - destroys the JWT cookie

image
image
image
image
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ your own extensions.
 
 It uses [React](https://reactjs.org/) and [TypeScript](https://www.typescriptlang.org/) for writing your extension, the [React Extension SDK](https://github.com/looker-open-source/extension-sdk-react) for interacting with Looker, and [Webpack](https://webpack.js.org/) for building your code.
 
+This version of the Kitchen sink requires Looker 7.9 or above.
```
Comment (jkaster) [inline README.md:8]: Users will never see 7.9. Do you intend this for 7.10?

File: env_config.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const dotenv = require('dotenv')
+
+module.exports = () => {
+  dotenv.config()
+  return Object.keys(process.env).reduce((accum, current) => {
+    accum[`process.env.${current}`] = JSON.stringify(process.env[current]);
+    return accum;
+  }, {});
+}
```
Comment (jkaster) [inline env_config.js:9]: What editor is inserting null characters at the end of the file?

File: server/index.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+// Log the user in.
+// If the client provides a google access token it will be validated.
+// Otherwise the user will be logged in anonymously.
+server.post('/auth', async (req, res) => {
+  const { type, access_token, expires_in, name, id } = req.body
+  if (type === 'Google') {
+    // For google login, verify the access token.
+    // If access token is not valid return a 401.
+    try {
+      const response = await axios.get(`https://oauth2.googleapis.com/tokeninfo?access_token=${access_token}`)
```
Comment (jkaster) [inline server/index.js:119]: FWIW if the extension SDK requires the Typescript SDK, this request could be fulfilled with a direct request from the transport layer (particularly the new `rawRequest` function) unless there's axios is being used for something other than HTTP requests. Reduce dependencies?

File: src/App.tsx [hunk 4]
```diff
@@ -60,7 +60,7 @@ export const App: React.FC<AppProps> = hot(() => {
   }
 
   return (
-    <ExtensionProvider onRouteChange={onRouteChange}>
+    <ExtensionProvider onRouteChange={onRouteChange} requiredLookerVersion=">=7.9.0">
```
Comment (jkaster) [inline src/App.tsx:63]: 7.10 for public release?

File: src/components/ExternalApiFunctions/components/GoogleSheetsDemo/GoogleSheetsDemo.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      if (GOOGLE_CLIENT_ID === '') {
+        updateErrorMessage(dataDispatch, 'Google client id has not been defined. Please see README.md for instructions.')
+      } else {
+        // Make sure the user has logged in using google (we need a google access token)
+        const { googleAccessToken } = location.state as any
+        if (!googleAccessToken || dataState.authOption !== AuthOption.Google) {
+          updateErrorMessage(dataDispatch, 'Login using Google to run the sheets demo')
+        } else {
+          // Read the spread sheet
+          const spreadsheetId = '1BxiMVs0XRA5nFMdKvBdBZjgmUUqptlbs74OgvE2upms'
```
Comment (jkaster) [inline src/components/ExternalApiFunctions/components/GoogleSheetsDemo/GoogleSheetsDemo.tsx:70]: Do your setup steps allow for anyone to access this spreadsheet id?

File: README.md (hunk 1 above)
Comment (bryans99) [reply README.md:8]: We have internal developers using it. I wanted to let them know. I can change it to 7.10 once 7.10 goes out the door

File: server/index.js (hunk 3 above)
Comment (bryans99) [reply server/index.js:119]: The above is my demo data server and goes nowhere near looker. Basically, the code is trying to point out that the access_token should not be trusted by itself in a server application. Further authorization needs to be done.

File: env_config.js (hunk 2 above)
Comment (bryans99) [reply env_config.js:9]: Not sure where you are seeing the null characters. The editor is visual code.

File: src/App.tsx (hunk 4 above)
Comment (bryans99) [reply src/App.tsx:63]: See previous comment. I can certainly update when 7.10 goes out the door

File: src/components/ExternalApiFunctions/components/GoogleSheetsDemo/GoogleSheetsDemo.tsx (hunk 5 above)
Comment (bryans99) [reply src/components/ExternalApiFunctions/components/GoogleSheetsDemo/GoogleSheetsDemo.tsx:70]: Yes. This is an example spreadsheet provided in googles examples. I should call that out though

File: env_config.js (hunk 2 above)
Comment (bryans99) [reply env_config.js]: Fixed. Not sure what happened

**Pull request merged**